src/gfx/convert.cpp

```diff
@@ -0,0 +1,866 @@
+/*
+ * This file is part of RGBDS.
+ *
+ * Copyright (c) 2022, Eldred Habert and RGBDS contributors.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "gfx/convert.hpp"
+
+#include <algorithm>
+#include <assert.h>
+#include <errno.h>
+#include <fstream>
+#include <inttypes.h>
+#include <memory>
+#include <optional>
+#include <png.h>
+#include <setjmp.h>
+#include <string.h>
+#include <tuple>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "defaultinitalloc.hpp"
+#include "helpers.h"
+
+#include "gfx/main.hpp"
+#include "gfx/pal_packing.hpp"
+#include "gfx/pal_sorting.hpp"
+#include "gfx/proto_palette.hpp"
+
+struct Rgba {
+	uint8_t red;
+	uint8_t green;
+	uint8_t blue;
+	uint8_t alpha;
+
+	Rgba(uint8_t r, uint8_t g, uint8_t b, uint8_t a) : red(r), green(g), blue(b), alpha(a) {}
+	Rgba(uint32_t rgba) : red(rgba), green(rgba >> 8), blue(rgba >> 16), alpha(rgba >> 24) {}
+
+	operator uint32_t() const {
+		auto shl = [](uint8_t val, unsigned shift) { return static_cast<uint32_t>(val) << shift; };
+		return shl(red, 0) | shl(green, 8) | shl(blue, 16) | shl(alpha, 24);
+	}
+	bool operator!=(Rgba const &other) const {
+		return static_cast<uint32_t>(*this) != static_cast<uint32_t>(other);
+	}
+
+	bool isGray() const { return red == green && green == blue; }
+
+	/**
+	 * CGB colors are RGB555, so we use bit 15 to signify that the color is transparent instead
+	 * Since the rest of the bits don't matter then, we return 0x8000 exactly.
+	 */
+	static constexpr uint16_t transparent = 0x8000;
+
+	static constexpr uint8_t transparency_threshold = 5; // TODO: adjust this
+	/**
+	 * Computes the equivalent CGB color, respects the color curve depending on options
+	 */
+	uint16_t cgbColor() const {
+		if (alpha < transparency_threshold)
+			return transparent;
+		if (options.useColorCurve) {
+			assert(!"TODO");
+		} else {
+			return (red >> 3) | (green >> 3) << 5 | (blue >> 3) << 10;
+		}
+	}
+};
+
+class ImagePalette {
+	// Use as many slots as there are CGB colors (plus transparency)
+	std::array<std::optional<Rgba>, 0x8001> _colors;
```

Comment:
```suggestion
	std::array<std::optional<Rgba>, 1 << (5 * 3) + 1> _colors;
```

Reply:
The calculation seems arbitrary at first glance, and has confused me at first. Constants would help, but I will hold off until some of them are (potentially) introduced by further modifications.

Reply:
Given that this appears in a couple other files, I still think it's worth a constant; or is it still too early to bother?

Reply:
I'm not sure what to name that constant, tbh.

Reply:
Bump?